### PR TITLE
Replace rest with graphql for theme level support

### PR DESCRIPTION
**Pull request by enmaboya**
The reason for switching from REST to GraphQL is the same as with access scopes. REST is deprecated. Additionally, REST is redundant; it retrieves too much data, making it roughly three times slower than GraphQL in this case. Moreover, I strongly doubt that the theme helper service was used by anyone or anywhere outside the package.

### CODE REVIEW

**Comment by Nikolow** [comment]
VerifyScopes middleware + ApiHelper from Services is also using REST API for activateCharge, createCharge and createUsageCharge

**Comment by Kyon147** [comment]
@enmaboya I am going to get this tested and deployed next week.

**Review by Kyon147 — DISMISSED** [review]

**Comment by enmaboya** [comment]
@Kyon147 @JonPurvis  Hi! I resolved the conflicts.

File: src/Actions/FetchMainTheme.php [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+use Illuminate\Support\Facades\Log;
+use Osiset\ShopifyApp\Contracts\ShopModel;
+
+final class FetchMainTheme
+{
+    /**
+     * @return array{id: ?string, name: ?string}
+     */
+    public function handle(ShopModel $shop): array
```
Comment (JonPurvis) [inline src/Actions/FetchMainTheme.php:15]: Were you aiming to handle `userErrors` here? From what I can see, the `themes` connection only exposes `edges`, `nodes`, and `pageInfo`:
https://shopify.dev/docs/api/admin-graphql/latest/connections/OnlineStoreThemeConnection

It's also not in the query selection set. Would it make more sense to check top-level `$response['errors']` and treat an empty `nodes` array as having no main theme?
https://shopify.dev/docs/api/admin-graphql/latest#status-and-error-codes

File: src/Actions/FetchThemeAssets.php [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+                        }
+                    }
+                }
+            }
+        }', [
+            'id' => $mainThemeId,
+            'filenames' => array_values($filenames),
+        ]);
+
+        if (blank(data_get($response['body']->toArray(), 'data.theme.userErrors'))) {
```
Comment (JonPurvis) [inline src/Actions/FetchThemeAssets.php]: Should this be checking `data.theme.files.userErrors` instead? Shopify's example puts `userErrors` on `files`, not on `theme` directly. Additionally, it needs to be in the query selection set:
https://shopify.dev/docs/api/admin-graphql/latest/queries/theme

Also, are you handling top-level `$response['errors']` anywhere (e.g. `ACCESS_DENIED`, `THROTTLED`)?
https://shopify.dev/docs/api/admin-graphql/latest#status-and-error-codes

File: src/Actions/FetchThemeAssets.php [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+            return array_map(fn (array $data) => [
+                'filename' => $data['filename'],
+                'content' => $data['body']['content'] ?? '',
```
Comment (JonPurvis) [inline src/Actions/FetchThemeAssets.php]: If the file body isn't `OnlineStoreThemeFileBodyText`, `content` won't be set. Does this mean that defaulting to `''` here would incorrectly mark a theme as UNSUPPORTED?

Shopify has three body types (`content`, `contentBase64`, `url`):
https://shopify.dev/docs/api/admin-graphql/latest/queries/theme
https://shopify.dev/docs/api/admin-graphql/latest/objects/OnlineStoreThemeFile

For template JSON we probably only expect text, should we skip or log when `content` is missing instead?

File: src/Actions/VerifyThemeSupport.php [hunk 4]
```diff
@@ -1,45 +1,119 @@
...
 use Osiset\ShopifyApp\Contracts\Queries\Shop as IShopQuery;
+use Osiset\ShopifyApp\Contracts\ShopModel;
 use Osiset\ShopifyApp\Objects\Enums\ThemeSupportLevel;
 use Osiset\ShopifyApp\Objects\Values\ShopId;
-use Osiset\ShopifyApp\Services\ThemeHelper;
 
-class VerifyThemeSupport
+final class VerifyThemeSupport
 {
+    private const ASSET_FILE_NAMES = ['templates/product.json', 'templates/collection.json', 'templates/index.json'];
```
Comment (JonPurvis) [inline src/Actions/VerifyThemeSupport.php:14]: These templates used to be configurable via `config('shopify-app.theme_support.templates')`. Was hardcoding product/collection/index intentional? Could we not build this from the config?

File: src/resources/config/shopify-app.php [hunk 5]
```diff
@@ -592,24 +592,11 @@
...
-         */
-        'templates' => ['product', 'collection', 'index'],
-        /*
-         * Interval for caching the request: minutes, seconds, hours, days, etc.
-         */
-        'cache_interval' => 'hours',
-        /*
-         * Cache duration
-         */
-        'cache_duration' => 12,
```
Comment (JonPurvis) [inline src/resources/config/shopify-app.php:612]: How come we're removing caching? Shopify recommend using caching

File: src/Actions/VerifyThemeSupport.php [hunk 6]
```diff
@@ -1,45 +1,119 @@
...
-                    return ThemeSupportLevel::FULL;
+        if (isset($mainTheme['id'])) {
+            /** @var array<int, array{filename: string, content: string}> */
+            $assets = $this->fetchThemeAssets->handle(
+                shop: $shop,
+                mainThemeId: $mainTheme['id'],
+                filenames: self::ASSET_FILE_NAMES
+            );
+            $templateMainSections = $this->mainSections(
+                shop: $shop,
```
Comment (JonPurvis) [inline src/Actions/VerifyThemeSupport.php:40]: If a shop is missing e.g. `templates/collection.json`, GraphQL returns fewer nodes, could that change FULL → PARTIAL → UNSUPPORTED compared to the old REST behaviour?

(The `@app` schema check itself looks right per Shopify: https://shopify.dev/docs/storefronts/themes/architecture/blocks/app-blocks)

Please could you add a test for a shop that only has `product.json`, so the expected outcome is clear?

File: src/Services/ThemeHelper.php [hunk 7]
```diff
@@ -1,245 +0,0 @@
...
-use Osiset\ShopifyApp\Contracts\Queries\Shop as IShopQuery;
-use Osiset\ShopifyApp\Contracts\ShopModel as IShopModel;
-use Osiset\ShopifyApp\Objects\Values\MainTheme;
-use Osiset\ShopifyApp\Objects\Values\ShopId;
-use Osiset\ShopifyApp\Util;
-
-/**
- * Helper for dealing with theme and theme assets.
- */
-class ThemeHelper
```
Comment (JonPurvis) [inline src/Services/ThemeHelper.php:16]: In case anyone was importing `ThemeHelper`, could you add changelog entries for their removal?

Same goes for `MainTheme`.

File: src/resources/config/shopify-app.php [hunk 8]
```diff
@@ -592,24 +592,11 @@
     | Checking theme compatibility
     |--------------------------------------------------------------------------
     |
-    | It is necessary to check if your application is compatible with
-    | the theme app blocks.
+    | It is necessary to check if your application is compatible with the theme app blocks.
     |
     */
 
     'theme_support' => [
```
Comment (JonPurvis) [inline src/resources/config/shopify-app.php:599]: Only `unacceptable_levels` is left under `theme_support` now. This means apps that customised `templates`, `cache_interval`, or `cache_duration` will lose that. Was that intentional?

File: tests/Actions/VerifyThemeSupportTest.php [hunk 9]
```diff
@@ -3,27 +3,21 @@
...
 
 use Osiset\ShopifyApp\Actions\VerifyThemeSupport;
-use Osiset\ShopifyApp\Contracts\Queries\Shop as IShopQuery;
 use Osiset\ShopifyApp\Objects\Enums\ThemeSupportLevel;
-use Osiset\ShopifyApp\Objects\Values\ShopId;
-use Osiset\ShopifyApp\Services\ThemeHelper;
+use Osiset\ShopifyApp\Test\Stubs\Api as ApiStub;
 use Osiset\ShopifyApp\Test\TestCase;
 
 class VerifyThemeSupportTest extends TestCase
```
Comment (JonPurvis) [inline tests/Actions/VerifyThemeSupportTest.php:10]: Would you be able to add tests for a few edge cases?

- GraphQL top-level `errors` (not just empty theme): https://shopify.dev/docs/api/admin-graphql/latest#status-and-error-codes
- Shop with only 1 of 3 template JSON files
- Template file body with JSON comments (the `json_validate` fallback)

File: .github/workflows/ci.yml [hunk 10]
```diff
@@ -50,9 +50,6 @@ jobs:
         if: matrix.validate == true
         run: composer validate --strict
 
-      - name: Install Laravel legacy factories support
-        if: matrix.laravel != '7.0'
-        run: composer require "laravel/legacy-factories:^1.1" --dev --no-interaction --no-update
```
Comment (JonPurvis) [inline .github/workflows/ci.yml:55]: Please can you do these changes in a separate branch and PR and keep this one just for the REST replacement.

File: .github/workflows/ci.yml [hunk 11]
```diff
@@ -82,7 +79,6 @@ jobs:
         run: vendor/bin/phpunit
 
       - name: Run static code analysis
-        if: matrix.analysis == true
```
Comment (JonPurvis) [inline .github/workflows/ci.yml:85]: Please can you do these changes in a separate branch and PR and keep this one just for the REST replacement.

File: phpstan.neon [hunk 12]
```diff
@@ -1,8 +1,12 @@
+includes:
+    - phpstan-baseline.neon
+
```
Comment (JonPurvis) [inline phpstan.neon]: Please can you do these changes in a separate branch and PR and keep this one just for the REST replacement.

File: phpstan.neon [hunk 13]
```diff
@@ -1,8 +1,12 @@
...
 parameters:
 	level: 1
 	paths:
 		- src
 		- tests
+	reportUnmatchedIgnoredErrors: false
```
Comment (JonPurvis) [inline phpstan.neon]: Please can you do these changes in a separate branch and PR and keep this one just for the REST replacement.

File: phpstan-baseline.neon [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+			message: '#^Deprecated in PHP 8\.4\: Parameter \#2 \$requestInstance \(Illuminate\\Http\\Request\) is implicitly nullable via default value null\.$#'
+			identifier: parameter.implicitlyNullable
+			count: 1
+			path: tests/TestCase.php
+
+		-
+			message: '#^Deprecated in PHP 8\.4\: Parameter \#3 \$cb \(Closure\) is implicitly nullable via default value null\.$#'
+			identifier: parameter.implicitlyNullable
+			count: 1
+			path: tests/TestCase.php
```
Comment (JonPurvis) [inline phpstan-baseline.neon]: Please can you do these changes in a separate branch and PR and keep this one just for the REST replacement.

**Review by JonPurvis — CHANGES_REQUESTED** [review]

**Comment by Kyon147** [comment]
@JonPurvis I agree with your comments, I will have a look and review.

File: src/Actions/FetchThemeAssets.php (hunk 2 above)
Comment (enmaboya) [reply src/Actions/FetchThemeAssets.php]: ok, I'll make the changes now.

File: src/Actions/FetchThemeAssets.php (hunk 3 above)
Comment (enmaboya) [reply src/Actions/FetchThemeAssets.php]: i agree, for json templates and .liquid sections, shopify always returns OnlineStoreThemeFileBodyText, so there won't be a bug on the current api behavior

but default " is probably wrong, that's why I changed the logic a bit.

File: src/Actions/VerifyThemeSupport.php (hunk 4 above)
Comment (enmaboya) [reply src/Actions/VerifyThemeSupport.php:14]: there is no point in this configuration, since only these templates are always checked.
this is written in the Shopify documentation.

 https://shopify.dev/docs/apps/build/online-store/verify-support
 
initially, I was the one who added this to the configuration (back in the osiset package), but it was unnecessary, I doubt that at least someone changed it.

File: src/resources/config/shopify-app.php (hunk 5 above)
Comment (enmaboya) [reply src/resources/config/shopify-app.php:612]: at the moment, VerifyThemeSupport is launched exactly once, at the time of installation.

further, if the user manually runs this action, it will be more correct if he caches the request himself, if he needs it.

also:
1) this was the only caching API call to the administrator, which made it the only place where we could provide outdated data.;
2) in the old approach, errors were also stored in the cache — an unsuccessful/empty REST response was cached for N hours

File: src/resources/config/shopify-app.php (hunk 8 above)
Comment (enmaboya) [reply src/resources/config/shopify-app.php:599]: yes, if the user wants to cache the result of VerifyThemeSupport, then they can do it in their application.

there is no point in caching the result of VerifyThemeSupport during installation, since it is run only once.

File: phpstan-baseline.neon (hunk 14 above)
Comment (enmaboya) [reply phpstan-baseline.neon]: done, I'll send the pull request later if necessary

File: phpstan.neon (hunk 13 above)
Comment (enmaboya) [reply phpstan.neon]: done, I'll send the pull request later if necessary

File: phpstan.neon (hunk 12 above)
Comment (enmaboya) [reply phpstan.neon]: done, I'll send the pull request later if necessary

File: .github/workflows/ci.yml (hunk 11 above)
Comment (enmaboya) [reply .github/workflows/ci.yml:85]: done, I'll send the pull request later if necessary

File: .github/workflows/ci.yml (hunk 10 above)
Comment (enmaboya) [reply .github/workflows/ci.yml:55]: done, I'll send the pull request later if necessary

File: src/Actions/FetchMainTheme.php (hunk 1 above)
Comment (enmaboya) [reply src/Actions/FetchMainTheme.php:15]: ok, I'll make the changes now.

File: src/Actions/VerifyThemeSupport.php (hunk 6 above)
Comment (enmaboya) [reply src/Actions/VerifyThemeSupport.php:40]: no, the fact that GraphQL returns fewer nodes does not change the classification compared to the old REST behavior.
the level of support is considered relative to those templates that actually exist, and not relative to a fixed list of 3 names ASSET_FILE_NAMES


tests added

File: tests/Actions/VerifyThemeSupportTest.php (hunk 9 above)
Comment (enmaboya) [reply tests/Actions/VerifyThemeSupportTest.php:10]: done, tests added

File: src/Services/ThemeHelper.php (hunk 7 above)
Comment (enmaboya) [reply src/Services/ThemeHelper.php:16]: I would love to, but where is the changelog? :)

**Comment by enmaboya** [comment]
@Kyon147 @JonPurvis i made edits and tried to respond to all comments.

File: tests/Actions/VerifyThemeSupportTest.php [hunk 15]
```diff
@@ -3,42 +3,96 @@
...
+        $action = $this->app->make(VerifyThemeSupport::class);
+
+        $result = call_user_func(
+            $action,
+            $shop->getId()
+        );
+
+        $this->assertNotNull($result);
+        $this->assertEquals(ThemeSupportLevel::FULL, $result);
+    }
```
Comment (JonPurvis) [inline tests/Actions/VerifyThemeSupportTest.php:55]: This test seems to be the exact same as `testStoreWithFullExtensionSupport`, which you've renamed from `testStoreWithUndefinedMainTheme` - can you remove this test OR tweak it to be what it was meant for? 

Once the above is sorted, I'll approve this PR.

**Review by JonPurvis — CHANGES_REQUESTED** [review]

File: tests/Actions/VerifyThemeSupportTest.php (hunk 15 above)
Comment (enmaboya) [reply tests/Actions/VerifyThemeSupportTest.php:55]: @JonPurvis added a check for multiple templates in "testStoreWithFullExtensionSupport"

**Review by JonPurvis — APPROVED** [review]
👍 

If @Kyon147 is happy with this implementation, then we can get it merged and tag a release.
